statistics/handle/handle.go

```diff
@@ -2047,3 +2009,73 @@ func (h *Handle) recordHistoricalStatsMeta(tableID int64, version uint64) error
 	}
 	return nil
 }
+
+// RecordHistoricalStatsToStorage records the given table's stats data to mysql.stats_history
+func (h *Handle) RecordHistoricalStatsToStorage(dbName string, tableInfo *model.TableInfo) (uint64, error) {
+	ctx := context.Background()
+	js, err := h.DumpStatsToJSON(dbName, tableInfo, nil)
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+	version := uint64(0)
+	for _, value := range js.Columns {
+		version = uint64(*value.StatsVer)
+		if version != 0 {
+			break
+		}
+	}
+	blocks, err := JSONTableToBlocks(js, maxColumnSize)
+	if err != nil {
+		return version, errors.Trace(err)
+	}
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	exec := h.mu.ctx.(sqlexec.SQLExecutor)
+	_, err = exec.ExecuteInternal(ctx, "begin pessimistic")
+	if err != nil {
+		return version, errors.Trace(err)
+	}
+	defer func() {
+		err = finishTransaction(ctx, exec, err)
+	}()
+	ts := time.Now().Format("2006-01-02 15:04:05.999999")
+
+	const sql = "INSERT INTO mysql.stats_history(table_id, stats_data, seq_no, version, create_time) VALUES (%?, %?, %?, %?, %?)"
+	for i := 0; i < len(blocks); i++ {
+		if _, err := exec.ExecuteInternal(ctx, sql, tableInfo.ID, blocks[i], i, version, ts); err != nil {
+			return version, errors.Trace(err)
+		}
+	}
+	return version, nil
+}
+
+// RecordAnalyzeJob inserts analyze job into mysql.analyze_jobs and gets job ID for further updating job.
+func (h *Handle) RecordAnalyzeJob(job *statistics.AnalyzeJob, procID uint64) error {
```

Comment:
Why does this method is under stats handle but other methods that manipulate `mysql.analyze_jobs` are in `executor/analyze.go`.

Reply:
We need to execute `insert into mysql.analyze_jobs ...` and `select last_insert_id()` in the same session to get the right ID. `RestrictedSQLExecutor` don't provide the method that executes multiple statements(It may execute the two statements using different sessions). I think that `baseAnalyzeExec.ctx` cannot execute the two statements either. Therefore, I choose `Handle.mu.ctx` to execute the two statements and put the function in `handle.go`.